### PR TITLE
Allow BIGINT to DECIMAL and DECIMAL to DOUBLE implicit conversions

### PR DESCRIPTION
**Pull request by arhimondr**

### CODE REVIEW

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java [hunk 1]
```diff
@@ -467,8 +467,8 @@ public void testShowColumnsPartitionKey()
         MaterializedResult expected = resultBuilder(getSession(), VARCHAR, VARCHAR, VARCHAR)
                 .row("grape", "bigint", "")
                 .row("orange", "bigint", "")
-                .row("pear", "varchar", "")
-                .row("apple", "varchar", "Partition Key")
+                .row("pear", "varchar(2147483647)", "")
```
Comment (losipiuk) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]: should not be needed after rebase to master

File: presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java [hunk 2]
```diff
@@ -104,6 +110,8 @@ public TypeRegistry(Set<Type> types)
         addType(JSON_PATH);
         addType(COLOR);
         addType(JSON);
+        addParametricType(VarcharParametricType.VARCHAR);
+        addParametricType(DecimalParametricType.DECIMAL);
         addParametricType(ROW);
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java]: Maybe better name would be `isSubType`. Though it is not exactly that. .... I do not know .... `typesMatch` is too generic name - at least give some comment with explanation.

File: presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java [hunk 3]
```diff
@@ -159,7 +167,9 @@ private Type instantiateParametricType(TypeSignature signature)
             return null;
         }
         Type instantiatedType = parametricType.createType(parameters);
-        checkState(instantiatedType.getTypeSignature().equals(signature), "Instantiated parametric type name (%s) does not match expected name (%s)", instantiatedType, signature);
+
+        // TODO: reimplement this check? Currently "varchar(Integer.MAX_VALUE)" fails with "varchar"
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java]: Put explicit precondition here that we expect `expectedTypeParameters.size()` to be either 0 or equal to size of `actualTypeParameters.size()`.

File: presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java [hunk 4]
```diff
@@ -216,7 +226,7 @@ private static boolean canCastTypeBase(String fromTypeBase, String toTypeBase)
         }
         switch (fromTypeBase) {
             case StandardTypes.BIGINT:
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java:227]: you should not change `parametersMatch` to `true` within the loop. Initial value should be set to `true` and then the loop should only change it to `false` if one of parameters does not match.

File: presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java [hunk 5]
```diff
@@ -216,7 +226,7 @@ private static boolean canCastTypeBase(String fromTypeBase, String toTypeBase)
...
-                return StandardTypes.DOUBLE.equals(toTypeBase);
+                return StandardTypes.DOUBLE.equals(toTypeBase) || StandardTypes.DECIMAL.equals(toTypeBase);
             case StandardTypes.DATE:
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java:229]: Why sometimes we return false and sometimes we throw IllegalStateException. This seem quite unnatural to me.

File: presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java [hunk 6]
```diff
@@ -248,7 +260,53 @@ public static boolean canCoerce(Type actualType, Type expectedType)
...
         Optional<TypeSignature> commonSuperTypeSignature = getCommonSuperTypeSignature(actualType, expectedType);
-        return commonSuperTypeSignature.isPresent() && commonSuperTypeSignature.get().equals(expectedType);
+        return commonSuperTypeSignature.isPresent() && typesMatch(commonSuperTypeSignature.get(), expectedType);
+    }
+
+    private static boolean typesMatch(TypeSignature actualType, TypeSignature expectedType)
+    {
+        if (!actualType.getBase().equals(expectedType.getBase())) {
+            return false;
+        }
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java:285]: rename `getCommonSuperTypeForLiteraTypes` to `getCommonSuperTypeForTypesWithLongParameters`

File: presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java [hunk 7]
```diff
@@ -349,4 +408,179 @@ else if (isCovariantParameterPosition(commonSuperTypeBase.get(), i)) {
...
+            return getCommonSuperTypeForVarchar(typeSignaturePair);
+        }
+
+        if (typeSignaturePair.is(StandardTypes.DECIMAL, UnknownType.NAME)) {
+            return getCommonSuperTypeForDecimals(new TypeSignaturePair(
+                    typeSignaturePair.get(StandardTypes.DECIMAL),
+                    createDecimalTypeSignature(1, 0)
+            ));
+        }
+
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java:468]: some comment what it means. Or generally describe logic of this method in 2-3 lines of comment at the beginning.

File: presto-main/src/main/java/com/facebook/presto/util/DecimalUtils.java [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+import com.facebook.presto.spi.type.TypeSignature;
+import com.facebook.presto.spi.type.TypeSignatureParameter;
+import com.facebook.presto.type.DecimalParametricType;
+import com.google.common.collect.ImmutableList;
+
+import java.math.BigInteger;
+
+import static com.facebook.presto.spi.type.LongDecimalType.MAX_DECIMAL_UNSCALED_VALUE;
+import static com.facebook.presto.spi.type.LongDecimalType.MIN_DECIMAL_UNSCALED_VALUE;
+import static com.facebook.presto.spi.type.StandardTypes.DECIMAL;
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/util/DecimalUtils.java]: This is already in `DecimalType`

File: presto-main/src/main/java/com/facebook/presto/util/DecimalUtils.java [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.StandardErrorCode;
+import com.facebook.presto.spi.type.LongDecimalType;
+import com.facebook.presto.spi.type.ShortDecimalType;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.spi.type.TypeParameter;
+import com.facebook.presto.spi.type.TypeSignature;
+import com.facebook.presto.spi.type.TypeSignatureParameter;
+import com.facebook.presto.type.DecimalParametricType;
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/util/DecimalUtils.java]: move it to `DecimalType`

File: presto-main/src/main/java/com/facebook/presto/util/DecimalUtils.java [hunk 10]
```diff
@@ -50,4 +57,16 @@ public static void checkOverflow(BigInteger value)
             throw new PrestoException(StandardErrorCode.INVALID_FUNCTION_ARGUMENT, "DECIMAL result exceeds 38 digits");
         }
     }
+
+    public static TypeSignature createDecimalTypeSignature(long precision, long scale)
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/util/DecimalUtils.java]: move it to DecimalType

File: presto-main/src/main/java/com/facebook/presto/util/DecimalUtils.java [hunk 11]
```diff
@@ -50,4 +57,16 @@ public static void checkOverflow(BigInteger value)
...
+    {
+        ImmutableList<TypeSignatureParameter> typeSignatureParameters =
+                ImmutableList.of(TypeSignatureParameter.of(precision), TypeSignatureParameter.of(scale));
+        return new TypeSignature(DECIMAL, typeSignatureParameters);
+    }
+
+    public static Type createDecimalType(long precision, long scale)
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/util/DecimalUtils.java]: this is already in DecimalType

File: presto-main/src/test/java/com/facebook/presto/metadata/TestSignature.java [hunk 12]
```diff
@@ -135,6 +140,62 @@ public void testBindUnknown()
     }
 
     @Test
+    public void testBindBigintToDecimal()
+            throws Exception
+    {
+        assertFunctionBind(
```
Comment (losipiuk) [inline presto-main/src/test/java/com/facebook/presto/metadata/TestSignature.java]: this assertions suck

File: presto-main/src/main/java/com/facebook/presto/metadata/Signature.java [hunk 13]
```diff
@@ -452,22 +451,37 @@ else if (canCoerce(boundParameters.get(parameter.getBase()), type) && typeParame
         // Bind parameter, if this is a free type parameter
         if (typeParameters.containsKey(parameter.getBase())) {
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/metadata/Signature.java]: break this method into smaller

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java [hunk 14]
```diff
@@ -5478,4 +5478,57 @@ public void testAccessControl()
...
+        assertQuery("SELECT CAST(292 AS BIGINT) + CAST(101 AS BIGINT)");
+        assertQuery("SELECT CAST(1.1 AS DECIMAL(38,1)) + CAST(292 AS BIGINT)");
+        assertQuery("SELECT CAST(1.1 AS DECIMAL(38,1)) + NULL");
+        assertQuery("SELECT CAST(1.1 AS DECIMAL(38,1)) IS NULL");
+        assertQuery("SELECT CAST(292 AS DECIMAL(38,1)) = CAST(292.1 AS DECIMAL(5,1))");
+        assertQuery("SELECT CAST(292 AS DECIMAL(38,1)) = 292");
+        assertQuery("SELECT CAST(292 AS DECIMAL(38,1)) = CAST(292 AS BIGINT)");
+
+        assertEqualsIgnoreOrder(
+                computeActual("SELECT ARRAY[CAST(282 AS DECIMAL(22,1)), CAST(282 AS DECIMAL(10,1))] || CAST(292 AS DECIMAL(5,1))"),
```
Comment (arhimondr) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: Add `ARRAY<DECIMAL> || LONG` test

File: presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java [hunk 15]
```diff
@@ -252,7 +260,53 @@ public static boolean canCoerce(Type actualType, Type expectedType)
     public static boolean canCoerce(TypeSignature actualType, TypeSignature expectedType)
     {
         Optional<TypeSignature> commonSuperTypeSignature = getCommonSuperTypeSignature(actualType, expectedType);
-        return commonSuperTypeSignature.isPresent() && commonSuperTypeSignature.get().equals(expectedType);
+        return commonSuperTypeSignature.isPresent() && typesMatch(commonSuperTypeSignature.get(), expectedType);
+    }
+
+    private static boolean typesMatch(TypeSignature actualType, TypeSignature expectedType)
```
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java]: This code looks similar to `com.facebook.presto.metadata.Signature#matchAndBindParameters` and `com.facebook.presto.type.TypeUtils#resolveCalculatedType`. Maybe this should be unified somehow? Maybe one pass over Signature that will "in place"/"online"/"live" checking for matching, coercions and simultaneously calculating literal parameters?

**Comment by losipiuk** [comment]
looks good - give it a try

File: presto-main/src/main/java/com/facebook/presto/type/TypeSignaturePair.java [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+    }
+
+    public TypeSignature get(String baseType)
+    {
+        return getSingleParameterByPredicate((type) -> type.getBase().equals(baseType));
+    }
+
+    public boolean bothTypesAreCalculated()
+    {
+        return typeList.stream().allMatch(TypeSignature::isCalculated);
```
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/type/TypeSignaturePair.java]: IMO this class would be simpler and easier to understand without this `List`
`return firstType.isCalculated() && secondType.isCalculated()`

File: presto-main/src/test/java/com/facebook/presto/type/TestTypeRegistry.java [hunk 17]
```diff
@@ -83,12 +91,38 @@ public void testCanCoerce()
...
+        assertFalse(TypeRegistry.canCoerce(parseTypeSignature("array(bigint)"), parseTypeSignature("array(decimal(2,1))")));
+
+        assertTrue(TypeRegistry.canCoerce(parseTypeSignature("decimal(22,1)"), DOUBLE.getTypeSignature()));
+        assertTrue(TypeRegistry.canCoerce(parseTypeSignature("decimal"), DOUBLE.getTypeSignature()));
+        assertTrue(TypeRegistry.canCoerce(decimalPS, DOUBLE.getTypeSignature()));
     }
 
-    @Test(expectedExceptions = IllegalArgumentException.class)
+    @Test
     public void testCanCoerceWithImplicitParameters()
```
Comment (pnowojski) [inline presto-main/src/test/java/com/facebook/presto/type/TestTypeRegistry.java]: remove this test

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java [hunk 18]
```diff
@@ -5478,4 +5478,57 @@ public void testAccessControl()
...
+     * It seems that once E is bound, we should bound it recursive in type arguments, similar to as we bind the base type
+     * for the signatures like do_something<E>(E,E):E
+     */
+    @Test(enabled = false)
+    public void testConcatenationWithNull()
+            throws Exception
+    {
+        assertQuery("SELECT 'something' || NULL");
+        assertEqualsIgnoreOrder(
+                computeActual("SELECT ARRAY[123, 123] || NULL"),
```
Comment (pnowojski) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java:5535]: decimal + double and double + bigint

**Comment by pnowojski** [comment]
except of this generally copy/paste of matchAndBind, canCoerce and resolveCalculatedTypes LGTM. It would be best to first push it as it is and later refactor

File: presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java (hunk 2 above)
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java]: this is not exactly `subType`. I agree that `typesMatch` is too generic. But i can't find any better name for now.

File: presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java (hunk 3 above)
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java]: I add the comment with explanation why i'm doing so.

File: presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java (hunk 4 above)
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java:227]: I'm going to extract the entire switch into separate method

File: presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java (hunk 5 above)
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java:229]: I add the comment with explanation why i'm doing so.

File: presto-main/src/main/java/com/facebook/presto/util/DecimalUtils.java (hunk 8 above)
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/util/DecimalUtils.java]: moved

Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/util/DecimalUtils.java]: I'm going to move createDecimalTypeSignature to DecimalType

File: presto-main/src/test/java/com/facebook/presto/metadata/TestSignature.java (hunk 12 above)
Comment (arhimondr) [inline presto-main/src/test/java/com/facebook/presto/metadata/TestSignature.java]: I know. It is quite hard to understand which parameter stands for what. But it is better to refactor it in a separate PR.

File: presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java (hunk 6 above)
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java:285]: done

File: presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java (hunk 7 above)
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java:468]: done

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java (hunk 18 above)
Comment (arhimondr) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java:5535]: done in a method above

File: presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java (hunk 15 above)
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java]: In order to make literal parameters to calculate on a fly we need to refactor all `matchAndBind` code. I think it is better to do as you said, as a separate PR.

**Comment by arhimondr** [comment]
All the comments are applied except one with `Signature` methods refactor.

**Pull request merged**